Language/Haskell/GhcMod/HomeModuleGraph.hs

```diff
@@ -206,17 +208,33 @@ updateHomeModuleGraph' env smp0 = do
            mapM_ go (Set.toList smp)
 
    step :: ModulePath -> m (Maybe (Set ModulePath))
-   step mp = runMaybeT $ do
-       (dflags, ppsrc_fn) <- MaybeT preprocess'
-       src <- liftIO $ readFile ppsrc_fn
-       imports mp src dflags
-    where
-      preprocess' :: m (Maybe (DynFlags, FilePath))
-      preprocess' = do
+   step mp = do
         let fn = mpPath mp
-        ep <- liftIO $ withLogger' env $ \setDf -> let
+        mapping <- lookupMMappedFile fn
+        runMaybeT $ do
+          let preprocess'' fn' = do
+                (dflags, ppsrc_fn) <- MaybeT (preprocess' fn')
+                src <- liftIO $ readFile ppsrc_fn
+                return (dflags, src)
+          (dflags, src) <- case mapping of
+            Just (RedirectedMapping to) -> preprocess'' to
+            Just (MemoryMapping (Just src)) -> do
+              (fp, hndl) <- liftIO $ getTemporaryDirectory >>= \tmpDir' ->
```

Comment:
Please re-use the tempdir in the `Cradle` type. You'll have to pass it down to this function but I don't want ghc-mod to create a hundred tempdirs per run.